### PR TITLE
Internal Application's Blog

### PR DESCRIPTION
**Pull request by ErikGmz**
# Description

- Added a blog that is exclusive for the application's users.
- Added posting system that is available for any user.
- Added posts commenting feature.

## Testing

- Run the rails db:migrate to properly create the database in a local PostgreSQL environment.
- Install Action Text and Active Storage.
- Run the rails db:seed command in order to load all the default users data.
- Run either bin/dev or rails server and open the localhost:3000 endpoint in any browser.
- Login with any account in order to check the blog.

### CODE REVIEW

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -1,3 +1,33 @@
...
+    def check_if_is_owner_or_admin
+      unless @post.user.id == current_user.id || current_user.is_an_admin?
+        redirect_to posts_path(page: 1), notice: "Only owners and admin are allowed to delete comments."
+      end
+    end
+
+    def comment_params
+      params.require(:comment).permit(:user_id, :post_id, :title, :rating, :content)
+    end
 end
```
Comment (PudinDeveloper) [inline app/controllers/comments_controller.rb:33]:
```suggestion
end

```

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -1,3 +1,75 @@
...
+    end
+
+    def post_params
+      params.require(:post).permit(:user_id, :title, :content)
+    end
+
+    def filtering_params(params)
+      params.slice(:users_ids, :title)
+    end
 end
```
Comment (PudinDeveloper) [inline app/controllers/posts_controller.rb:75]:
```suggestion
end

```

File: app/controllers/reviews_controller.rb [hunk 3]
```diff
@@ -21,6 +28,14 @@ def check_if_is_guest_or_admin
...
+      review = @accommodation.reviews.find(params[:id])
+
+      unless review.user.id == current_user.id || current_user.is_an_admin?
+        redirect_to accommodations_path(page: 1), notice: "Only owners and admin are allowed to delete reviews."
+      end
+    end
+
     def review_params
       params.require(:review).permit(:user_id, :accommodation_id, :title, :rating, :content)
     end
```
Comment (PudinDeveloper) [inline app/controllers/reviews_controller.rb:41]:
```suggestion
    end
    
```

File: app/models/post.rb [hunk 4]
```diff
@@ -6,4 +6,7 @@ class Post < ApplicationRecord
 
   validates :user_id, presence: true
   validates :title, presence: true, length: { maximum: 100 }
+
+  scope :filter_by_users_ids, ->(users_ids) { where(user_id: users_ids).order(updated_at: :desc) }
+  scope :filter_by_title, ->(title) { where("title LIKE ?", "%#{title}%").order(updated_at: :desc) }
 end
```
Comment (PudinDeveloper) [inline app/models/post.rb:12]:
```suggestion
end

```

File: app/views/comments/_new.html.erb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+  <div class="my-5">
+    <%= form.label :content %>
+    <%= form.rich_text_area :content, class: "block shadow rounded-md border border-gray-200 outline-none px-3 py-2 mt-2 w-full" %>
+  </div>
+
+  <div class="inline">
+    <%= form.submit class: "rounded-lg py-3 px-5 bg-blue-600 text-white inline-block font-medium cursor-pointer" %>
+  </div>
+<% end %>
```
Comment (PudinDeveloper) [inline app/views/comments/_new.html.erb:17]:
```suggestion
<% end %>

```

File: app/views/home/index.html.erb [hunk 6]
```diff
@@ -14,6 +14,7 @@
         <%= link_to 'Users', users_management_index_path %>
       <% end %>
 
+      <%= link_to 'Blog', posts_path %>
       <%= link_to 'Profile', profile_path %>
       <%= link_to 'Log out', destroy_user_session_path, data: { turbo_method: :delete } %>
     <% end %>
```
Comment (PudinDeveloper) [inline app/views/home/index.html.erb:20]:
```suggestion
    <% end %>
    
```

File: app/views/posts/_comments.html.erb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<h1 class="font-bold text-4xl">Comments</h1>
+
+<div id="comments" class="min-w-full">
+  <%= render "comments/new", post: post %>
+  <%= render comments %>
+</div>
+
+<%== pagy_nav(@pagy) if @pagy.pages > 1 %>
```
Comment (PudinDeveloper) [inline app/views/posts/_comments.html.erb:8]:
```suggestion
<%== pagy_nav(@pagy) if @pagy.pages > 1 %>

```

File: app/views/posts/_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+  <div class="my-5">
+    <%= form.label :content %>
+    <%= form.rich_text_area :content, class: "block shadow rounded-md border border-gray-200 outline-none px-3 py-2 mt-2 w-full" %>
+  </div>
+
+  <div class="inline">
+    <%= form.submit class: "rounded-lg py-3 px-5 bg-blue-600 text-white inline-block font-medium cursor-pointer" %>
+  </div>
+<% end %>
```
Comment (PudinDeveloper) [inline app/views/posts/_form.html.erb:29]:
```suggestion
<% end %>

```

File: app/views/posts/_post.html.erb [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    <strong class="block font-medium mb-1">Content:</strong>
+    <%= post.content %>
+  </p>
+
+  <% if action_name != "show" %>
+    <%= link_to "Show this post", post, class: "rounded-lg py-3 px-5 bg-gray-100 inline-block font-medium" %>
+    <%= link_to "Edit this post", edit_post_path(post), class: "rounded-lg py-3 ml-2 px-5 bg-gray-100 inline-block font-medium" %>
+    <hr class="mt-6">
+  <% end %>
+<% end %>
```
Comment (PudinDeveloper) [inline app/views/posts/_post.html.erb:19]:
```suggestion
<% end %>

```

File: app/views/posts/_search_filter_form.html.erb [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+                            class: "block shadow rounded-md border border-gray-200 outline-none px-3 py-2 mt-2 w-72" %>
+      </div>
+    </div>
+
+    <div class='flex flex-col items-center min-[250px]:flex-row'>
+      <%= form.submit 'Search', class: 'rounded-lg py-3 px-5 bg-gray-100 inline-block font-medium' %>
+      <%= link_to 'Clear', posts_path, class: 'rounded-lg py-3 px-5 bg-gray-100 inline-block font-medium ml-3 no-underline' %>
+    </div>
+  </div>
+<% end %>
```
Comment (PudinDeveloper) [inline app/views/posts/_search_filter_form.html.erb:24]:
```suggestion
<% end %>

```

File: app/views/posts/edit.html.erb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+<div class="mx-auto md:w-2/3 w-full">
+  <h1 class="font-bold text-4xl">Editing post</h1>
+
+  <%= render "form", post: @post %>
+
+  <%= link_to "Show this post", @post, class: "ml-2 rounded-lg py-3 px-5 bg-gray-100 inline-block font-medium" %>
+  <%= link_to "Back to posts", posts_path, class: "ml-2 rounded-lg py-3 px-5 bg-gray-100 inline-block font-medium" %>
+</div>
```
Comment (PudinDeveloper) [inline app/views/posts/edit.html.erb:8]:
```suggestion
</div>

```

File: app/views/posts/index.html.erb [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+      <%= link_to "Show this post", post, class: "rounded-lg py-3 px-5 bg-gray-100 inline-block font-medium" %>
+      <% if post.user.id == current_user.id || current_user.is_an_admin? %>
+        <%= link_to "Edit this post", edit_post_path(post), class: "rounded-lg py-3 ml-2 px-5 bg-gray-100 inline-block font-medium" %>
+      <% end %>
+      <hr class="mt-6">
+    <% end %>
+  </div>
+
+  <%== pagy_nav(@pagy) if @pagy.pages > 1 %>
+</div>
```
Comment (PudinDeveloper) [inline app/views/posts/index.html.erb:30]:
```suggestion
</div>

```

File: app/views/posts/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+<div class="mx-auto md:w-2/3 w-full">
+  <h1 class="font-bold text-4xl">New post</h1>
+
+  <%= render "form", post: @post %>
+
+  <%= link_to "Back to posts", posts_path, class: "ml-2 rounded-lg py-3 px-5 bg-gray-100 inline-block font-medium" %>
+</div>
```
Comment (PudinDeveloper) [inline app/views/posts/new.html.erb:7]:
```suggestion
</div>

```

File: app/views/posts/show.html.erb [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+        <div class="inline-block ml-2">
+          <%= button_to "Destroy this post", post_path(@post), method: :delete, class: "mt-2 rounded-lg py-3 px-5 bg-gray-100 font-medium" %>
+        </div>
+      <% end %>
+      <%= link_to "Back to posts", posts_path, class: "ml-2 rounded-lg py-3 px-5 bg-gray-100 inline-block font-medium" %>
+    </div>
+
+    <%= render "posts/comments", post: @post, comments: @comments %>
+  </div>
+</div>
```
Comment (PudinDeveloper) [inline app/views/posts/show.html.erb:21]:
```suggestion
</div>

```

File: app/views/reviews/_review.html.erb [hunk 15]
```diff
@@ -19,6 +19,10 @@
     <%= review.content %>
   </p>
 
+  <% if review.user.id == current_user.id || current_user.is_an_admin? %>
+    <%= button_to "Destroy this review", [review.accommodation, review], method: :delete, class: "mt-2 rounded-lg py-3 px-5 bg-gray-100 font-medium" %>
+  <% end %>
+
   <% if action_name != "show" %>
     <hr class="mt-6">
   <% end %>
```
Comment (PudinDeveloper) [inline app/views/reviews/_review.html.erb:28]:
```suggestion
  <% end %>
  
```

File: db/seeds.rb [hunk 16]
```diff
@@ -60,3 +60,18 @@
...
+  content = ActionText::RichText.where(record_type: 'Post', record_id: post.id,
+                                     name: 'content').first_or_initialize
+  content.update(body: Faker::Lorem.sentence)
+
+  comment = Comment.where(title: "Comment no. #{post.id}", post_id: post.id).first_or_initialize
+  comment.update(user_id: guest_id)
+  content = ActionText::RichText.where(record_type: 'Comment', record_id: comment.id,
+                                     name: 'content').first_or_initialize
+  content.update(body: Faker::Lorem.sentence)
+end
```
Comment (PudinDeveloper) [inline db/seeds.rb:77]:
```suggestion
end

```

**Review by PudinDeveloper — APPROVED** [review]
Suggestion to a lot of files: Add a line at the end of each file

**Review by MauricioNC — APPROVED** [review]

**Review by STR1ND3R — APPROVED** [review]

**Review by PudinDeveloper — APPROVED** [review]

**Comment by PudinDeveloper** [comment]
Solve branch conflicts :)

**Pull request merged**